aeternity/aens.py

```diff
@@ -0,0 +1,223 @@
+    def __init__(self, *args, payload=None):
+        super().__init__(*args)
+        self.payload = payload
+
+    def __str__(self):
+        return super().__str__() + '\npayload\n' + str(self.payload)
+
+class MissingPreclaim(AENSException):
+    pass
+
+
+class PreclaimFailed(AENSException):
+    pass
+
+
+class TooEarlyClaim(AENSException):
+    pass
+
+
+class ClaimFailed(AENSException):
+    pass
+
+
+class NameNotAvailable(AENSException):
+    pass
+
+
+class UpdateError(Exception):
+    pass
+
+
+class NameStatus:
+    REVOKED = 'REVOKED'
+    UNKNOWN = 'UNKNOWN'
+    AVAILABLE = 'AVAILABLE'
+    PRECLAIMED = 'PRECLAIMED'
+    CLAIMED = 'CLAIMED'
+    TRANSFERRED = 'TRANSFERRED'
+
+
+class Name:
+    Status = NameStatus
+
+    def __init__(self, domain, client=None):
+        if client is None:
+            client = EpochClient()
+        self.__class__.validate_name(domain)
+
+        self.client = client
+        self.domain = domain
+        self.status = NameStatus.UNKNOWN
+        # set after preclaimed:
+        self.preclaimed_block_height = None
+        self.preclaimed_commitment_hash = None
+        self.preclaim_salt = None
+        # set after claimed
+        self.name_hash = None
+        self.name_ttl = 0
+        self.pointers = []
+
+    @classmethod
+    def validate_pointer(cls, pointer):
+        if not pointer.startswith(('ak$')):
+            raise ValueError('pointer addresses must start with in ak$')
+
+    @classmethod
+    def validate_name(cls, domain):
+        # TODO: validate according to the spec!
+        # TODO: https://github.com/aeternity/protocol/blob/master/AENS.md#name
+        if not domain.endswith(('.aet', '.test')):
+            raise InvalidName('AENS TLDs must end in .aet')
+
+    def update_status(self):
+        response = self.client.local_http_get('name', params={'name': self.domain})
+        wasnt_found = response.get('reason') == 'Name not found'
+        # e.g. if the status is already PRECLAIMED or CLAIMED, don't reset
+        # it to AVAILABLE.
+        if wasnt_found and self.status == NameStatus.UNKNOWN:
+            self.status = NameStatus.AVAILABLE
+        else:
+            self.status = NameStatus.CLAIMED
+            self.name_hash = response['name_hash']
+            self.name_ttl = response['name_ttl']
+            self.pointers = json.loads(response['pointers'])
+
+    def is_available(self):
+        self.update_status()
+        return self.status == NameStatus.AVAILABLE
+
+    def check_claimed(self):
+        self.update_status()
+        return self.status == NameStatus.CLAIMED
+
+    def full_claim_blocking(self):
+        if not self.is_available():
+            raise NameNotAvailable(self.domain)
+        self.preclaim()
+        self.claim_blocking()
+
+    def preclaim(self):
+        # check which block we used to create the preclaim
+        self.preclaimed_block_height = self.client.get_top_block()
+        self.preclaim_salt = random.randint(0, 2**64)
```

Comment:
On the one hand I think it's good that you overtake the users decision to chose a salt by himself which would probably end up with users hard coding the salt (worst of all cases). At the same time, I am not fully convinced that the preclaim_salt should be part of the Name's state.

Reply:
Since I need the salt in the next operation `claim` again I think it's correct to do this. I guess I could mark it as private e.g. use `_preclaim_salt`